### ISSUE TITLE
Italic ligatures not working as expected

### ISSUE THREAD

**Issue by mskelton** [kiliman/operator-mono-lig#104]
Maybe I just did something wrong, but the italic ligatures aren't showing up for me in VS Code or IntelliJ.  Below is a screenshot showing what I'm seeing.

![2019-08-07_22-02-10](https://user-images.githubusercontent.com/25914066/62672179-3f819780-b95f-11e9-8059-43bd6f872243.png)

Without the comments, they show just fine.

![2019-08-07_22-04-13](https://user-images.githubusercontent.com/25914066/62672186-4d371d00-b95f-11e9-85f6-7acee9e891b7.png)

This is the case for all weights.

**Comment by markypython** [comment]
If I remove the letters in the comments, then they start to work.

![2019-08-07_22-07-45](https://user-images.githubusercontent.com/25914066/62672298-cc2c5580-b95f-11e9-93c7-263041f22409.png)

🤷‍♂

**Comment by Kryptonov** [comment]
Seeing the same here in VIM. Regular ligatures are fine, Italics seem "borky", sometimes they behave irratic, sometimes they don't show up at all.
Thanks for your work, btw, and of course to Kiliman.

Using:
VIM - Vi IMproved 8.1 (2018 May 18, compiled Aug  3 2019 18:05:22)
macOS version (10.14.6 (18G87))
Including 'patches': 1-1800

**Comment by kiliman** [comment]
Yeah, I've noticed it too. The odd thing is that the ligatures work, but it depends on the context.

For example, here's VS Code. Notice in the editor, the ligatures break when mixed with letters. But in Polacode (literally side-by-side) the ligatures show up. I believe it's because Polacode uses a web view to render.

![image](https://user-images.githubusercontent.com/47168/62944677-37848600-bdab-11e9-9208-bbae5a05eff3.png)

Here's the sample.js file viewed on GitHub using Chrome. Notice how the italic ligatures work there as well. Don't mind the regular ligatures. That's just GitHub not tokenizing correctly since some of those are not valid Javascript.

![image](https://user-images.githubusercontent.com/47168/62944749-5d118f80-bdab-11e9-8c63-122323c2476d.png)

And finally, I went and changed the font to always be Italic and this is what it looks like in Code. Notice how the non-commented text is displaying the italic ligatures just fine. But it's still broken inside the comments.

![image](https://user-images.githubusercontent.com/47168/62945082-05bfef00-bdac-11e9-8bf3-8a0d044098af.png)

Anyway, I'm not saying there is NO issue with the font/ligatures. It's hard to say what is causing it.

Fira Code displays the "italics" just fine, because it doesn't have true italics, so the OS is giving VS Code a synthetic italic.

I've also tried Victor Mono (includes ligatures and true italics), and it works fine, however, they are using double and triple width ligatures like my v1.0 font, so VS Code only sees one glyph, not the 2-3 glyphs I use. So it's not really comparing apples to apples.

I haven't tried Dank Mono (also has ligatures and true italics). I don't know if they use the "liga" method (v1) or "calt" method (v2).

This has been a low priority because I rarely comment out big blocks of code, so italic ligatures aren't that important to me. If anyone is willing to do the research, I'll be willing to try and fix it.

**Comment by kiliman** [comment]
There is a workaround to get italic ligatures to show up properly.

Add the following setting:  `"editor.renderWhitespace": "all",`

The drawback is that you now have all those "whitespace" dots. I'm not sure why VS Code renders text differently here.

![image](https://user-images.githubusercontent.com/47168/66148890-eab97080-e5df-11e9-85a6-ac31a34e7b28.png)

I wonder if there's a setting to specify which glyph to use for the whitespace dot. If so, we can just replace it with a space!

**Comment by kiliman** [comment]
Figured out a work around.

As I stated above, the italic ligatures display properly if you set `"editor.renderWhiteSpace": "all"`

However, that results in lots of dots being displayed on the screen.

It turns out you can set the color of the "whitespace" dots. 

Add this setting:

```json
  "editor.renderWhitespace": "all",
  "workbench.colorCustomizations": {
    "editorWhitespace.foreground": "#FFFFFF00"
  },
```

The color is RGBA. By setting the alpha to 0, we make the dots invisible. Voila! It works.

There's still some quirks as you can see in the screenshot above. Note sure why that line is rendered in red, but unfortunately the whitespace dots still show up there.

Anyway, unless anyone else figures out why this isn't working properly, I'm going to consider this issue "closed". I'll leave it open a while longer.

![image](https://user-images.githubusercontent.com/47168/66170160-291b5380-e611-11e9-88c8-c83b1d4c590d.png)

**Comment by kiliman** [comment]
I figured out why the comment is in red above. I have the *Better Comments*[1] extension and one of the features is that if you start your comment with a `!` it will highlight the comment in red.

The first ligature in the line above is the `!=` ligature. That is why it's highlighted in red. Still doesn't explain why whitespace rendering is busted (but that's probably due to the extension doing its own rendering).

Anyway, mystery solved!

[1] https://github.com/aaron-bond/better-comments

**Comment by kufii** [comment]
unfortunately this workaround doesn't really work for me because I like to set render whitespace to `selection` (and it's actually a feature I use very frequently)

**Comment by Kryptonov** [comment]
Also, this seems to be VSCode specific, the issue is much wider than that. Unfortunately I know little to nothing about type and fonts so I can only comment, which is a bit frustrating.

**Comment by kiliman** [comment]
Yes, I agree this isn't really a "solution". Unfortunately I'm not sure what's causing it. The ligatures do show up under certain circumstances, so I don't believe it's an issue with the font itself.

**Comment by kufii** [comment]
it is really weird and I have no idea why it would be happening, but I tried it out in several different applications (Chrome, Firefox, iTerm) and in all of them the ligatures don't display correctly in italics. That makes me think it's some obscure issue with the font itself

**Comment by cybergrunge** [comment]
I can confirm ligatures in italic (light) don't work anywhere in the system (macOS).

**Comment by kiliman** [comment]
@cybergrunge I'm not sure what the issue is for you, but I just sent my default font to Light Italic, and VS Code displays them just fine. macOS Catalina

![image](https://user-images.githubusercontent.com/47168/67100204-401f8100-f18d-11e9-91a5-d5648ece5b23.png)

Email me kiliman@gmail.com and I'll send you a test font.

**Comment by cybergrunge** [comment]
In Sublime Text on macOS you'll have to write the condensed name of the font in order to have italic ligatures.

Set  `"font_face": "OperatorMonoLig-Light"
`
Don't set `"font_face": "Operator Mono Lig Light"
`

**Comment by amitkparekh** [comment]
I'm having issues with the ligatures outside of VSCode, I have a feeling its the font file as no other font is having the issue, but I have no idea why.

**Comment by kiliman** [comment]
I think I figured out what's going on with the Italic fonts. Apparently there's an issue with the GSUB tables in the original font. Since I append my `calt` tables after, there must be some kind of conflict, and my tables are being ignored.

As a test of my hypothesis, I edited the TTX file of the original and removed the existing GSUB tables, then ran my build and sure enough, the ligatures came out perfectly.

I'm going to see if I can adjust the table ordering before going with the drastic option of removing the existing tables. Anyway, it looks like a solution if close at hand.

![image](https://user-images.githubusercontent.com/47168/68484924-49d85980-020c-11ea-9765-c282f03c8d00.png)

**Comment by amitkparekh** [comment]
@kiliman Have you managed to make any progress towards a possible solution?

**Comment by kiliman** [comment]
Is it just me or are italic ligatures finally working? I just updated to v1.45 (April 2020) version and noticed it said syntax highlighting was improved, so I figured I'd check to see if that fixed the italics and sure enough it does (at least for me)

![image](https://user-images.githubusercontent.com/47168/81621335-2d37fc80-93bc-11ea-9519-baae69d5184e.png)

Anyway, please check in and let me know if you are still having issues.

Thanks!

**Comment by mskelton** [comment]
Nope, still broken for me.  I believe it has something to do with the variable names

```js
// a == b
// a === b
// a <= b
```

Even more interesting is that when I highlight the selection it works but when it isn't highlighted it doesn't 😕

**Comment by kiliman** [comment]
Ha, you'll need to post a screenshot, because I have updated GitHub to use the ligature font, so I actually see it properly (using my Cobalt2 theme)

![image](https://user-images.githubusercontent.com/47168/81690868-eaad0900-9429-11ea-9684-94ba7267c2d6.png)

Anyway, I think I realized what it is. I had patched the original TTX file and removed the `GSUB` node, so the only thing added was the `calt` tables for the ligature substitution. That seems to be the trick. It looks like something in the original tables are causing a conflict with the ligatures.

I'm going to create a flag to `--italics-hack` so we can test it before making it the default.

**Comment by kiliman** [comment]
Ugh... I added the hack and regenerated the font and now it stopped working 😭

I'll have to compare with the original font I used.

Back to the drawing board.

**Comment by kiliman** [comment]
Ok, I think I figured out the problem. It's now working for me. I was applying the hack in the `buildGsubTables()` function, which was called per ligature. So it was wiping out the tables that I was creating. I moved the hack to `initGsubTables()` and it appears to be working.

Try checking out the `feature/italics-hack` branch and run `./build.sh --italics-hack` then reinstall the fonts and let me know if it's working for you.

Thanks!

**Comment by kiliman** [comment]
You can also build for just a single font by running:

`./build.sh --italics-hack OperatorMonoSSmLig-BookItalic`

**Comment by mskelton** [comment]
I'll give this a try tonight

**Comment by kiliman** [comment]
Another weird thing.
 
Works fine non-italic, but drops the ! in italics. 🤷‍♂️

![image](https://user-images.githubusercontent.com/47168/81720914-08d83080-944d-11ea-8ba5-7692cbeeb6ed.png)

**Comment by mskelton** [comment]
It still is broken for me.  I ran the build command you sent and installed the font but it still didn't fix the italics.

**Comment by amitkparekh** [comment]
There might be [font cache issues](https://glyphsapp.com/tutorials/eliminating-font-cache-problems) causing these differences? Glyphs recommends using their [TextPreview](https://glyphsapp.com/tools/textpreview) tool as a way of testing the fonts without running into these issues. 

Is it still working/not working if you put a freshly built version of the fonts into TextPreview?

**Comment by kiliman** [comment]
I sent you both a copy of the working italic. Let me know if it works for you.

**Comment by amitkparekh** [comment]
I loaded it up in TextPreview to see how it looks; entire screenshot of it is below. It appears that some of the ligatures are not showing up? I can't see `www`, `<!--`, but the arrows and equals are now starting to show. 

<img width="665" alt="Screenshot 2020-05-13 at 08 24 55" src="https://user-images.githubusercontent.com/7276308/81783477-48595800-94f3-11ea-9587-0c7e057477a7.png">

**Comment by kiliman** [comment]
Yeah, I'm not sure what makes the italic version different. But at least the bulk of them work.

Thanks for testing. I'm going to go ahead and merge my hack into master.

**Comment by kufii** [comment]
I tried rebuilding with the latest version and the italics ligatures still don't work for me

**Comment by mskelton** [comment]
For anyone following this issue, it has been resolved in version [2.4.0](https://github.com/kiliman/operator-mono-lig/releases/tag/v2.4.0).
